examples/configmapgen_controller.rs

```diff
@@ -112,16 +116,45 @@ async fn main() -> Result<()> {
     let cmgs = Api::<ConfigMapGenerator>::all(client.clone());
     let cms = Api::<ConfigMap>::all(client.clone());
 
-    Controller::new(cmgs, ListParams::default())
-        .owns(cms, ListParams::default())
-        .run(reconcile, error_policy, Context::new(Data { client }))
-        .for_each(|res| async move {
-            match res {
-                Ok(o) => info!("reconciled {:?}", o),
-                Err(e) => warn!("reconcile failed: {}", Report::from(e)),
-            }
-        })
-        .await;
+    log::info!("starting configmapgen-controller");
+    log::info!("press <enter> to force a reconciliation of all objects");
+    log::info!("press <ctrl>+c to shut down gracefully");
+
+    let (mut reload_tx, reload_rx) = futures::channel::mpsc::channel(0);
+    // Using a regular background thread since tokio::io::stdin() doesn't allow aborting reads,
+    // and its worker prevents the Tokio runtime from shutting down.
+    std::thread::spawn(move || {
+        for _ in std::io::BufReader::new(std::io::stdin()).lines() {
+            let _ = reload_tx.try_send(());
+        }
+    });
+
+    let (graceful_shutdown_tx, graceful_shutdown_rx) = futures::channel::oneshot::channel();
+    let forceful_shutdown = async {
+        tokio::signal::ctrl_c().await.unwrap();
+        graceful_shutdown_tx.send(()).unwrap();
+        log::info!("starting graceful shutdown, press <ctrl>+c again to shut down forcefully");
+        tokio::signal::ctrl_c().await.unwrap();
+        log::info!("forceful shutdown requested");
+        std::process::exit(0);
+    };
 
+    futures::future::select(
+        Controller::new(cmgs, ListParams::default())
+            .owns(cms, ListParams::default())
+            .reconcile_all_on(reload_rx.map(|_| ()))
+            .graceful_shutdown_on(graceful_shutdown_rx.map(|_| ()))
+            .run(reconcile, error_policy, Context::new(Data { client }))
+            .for_each(|res| async move {
+                match res {
+                    Ok(o) => info!("reconciled {:?}", o),
+                    Err(e) => warn!("reconcile failed: {}", Report::from(e)),
+                }
+            })
+            .boxed(),
+        forceful_shutdown.boxed(),
```

Comment:
although you do seem to need a select on the controller future vs. the shutdown future. That feels a bit subtle. Do you not want the controller to complete its outstanding items?

Reply:
That's the difference. The first `ctrl+c` initiates the graceful shutdown (by resolving `graceful_shutdown_rx`), the second means that we just want the process to die die die ASAP.

Reply:
Do we need to codify the latter into the app? Kubernetes will give us `SIGTERM`, wait 30s, then send a `SIGKILL`. We're don't really get to do anything after the `SIGKILL`, so it might not be worth trying to handle it.

Reply:
When running in-cluster, you're correct (though that actually raises the problem that iirc the example only listens for SIGINT, not SIGTERM).

When running locally, well, it's annoying to have to switch tab and `pkill`.

Reply:
right, we don't have a way of propagating a forceful termination within the applier. got it. so for the example we need the full setup.

Reply:
wait, actually, do we need the `forceful_shutdown` future? the async block that creates it is doing `process::exit` on the second `ctrl-c`? wouldn't that just stop everything?

Reply:
> right, we don't have a way of propagating a forceful termination within the applier. got it. so for the example we need the full setup.

Yeah, the whole point of the forceful termination would be to bypass everything and burn it to the ground.

> wait, actually, do we need the `forceful_shutdown` future? the async block that creates it is doing `process::exit` on the second `ctrl-c`? wouldn't that just stop everything?

We need something that waits for waits for the second signal, but doesn't trigger for the first. That means that we need to keep them in the same "path", so to speak.

The `std::process::exit` is also not strictly necessary, there's a difference between forceful and forceful. Essentially, you could say that we have six potential "levels" of grace that we could potentially implement:

1. The [Kronblom](http://annorlundaunderbara.com/wp-content/uploads/2015/06/Unknown3.jpeg) shutdown: when we initiate a shutdown, stop taking new scheduling requests, but let all currently *scheduled* reconciliations run and finish
   - This is what `scheduler` currently implements in `master` (before this PR), but due to `applier`'s circular nature this isn't actually usable in `applier` anyway (since it doesn't have the cutoff that this PR implements)
   - Depending on whether we still allow retries to be scheduled, this may never terminate
2. The [slightly overcooked](https://cdn-attachments.timesofmalta.com/92d8a06762be5aa090bcbb13d341c35b8699ec78-1485165073-5885d211-620x348.jpg) shutdown: like the above, but only let currently *pending* reconciliations run (that is, they have already expired, but haven't started yet for whatever reason) while dropping reconciliations that are scheduled into the future
3. The [graceful](https://www.wikihow.com/images/thumb/7/73/Be-Graceful-Step-2-Version-3.jpg/aid192280-v4-728px-Be-Graceful-Step-2-Version-3.jpg) shutdown: wait for all running reconciliations to finish, but do not start any new ones
   - This is what this PR calls a "graceful" shutdown
4. The [forceful](https://media.istockphoto.com/photos/forceful-businessman-boxing-picture-id1214489290?k=6&m=1214489290&s=612x612&w=0&h=FwL-bjAZ-jL0tu9LzPP4DYnGmiM83HN_WWVJibVGD3g=) shutdown: abort all currently running reconciliations, but wait for them to cancel orderly (essentially: wait for them to hit the next `.await`)
5. The [Brütal](https://www.spielejournalist.de/wp-content/uploads/2020/05/bruetal-legend-jack-black.jpg) shutdown: `std::process::exit`
   - This is what the example calls a "forceful" shutdown
6. The [Spın̈al](http://www.laut.de/Spinal-Tap/spinal-tap-189373.jpg) shutdown: you didn't need that computer anyway, did you?

From this list, this PR adds support for the graceful shutdown, while the forceful, Brütal, and Spın̈al shutdowns were already supported (kind of unavoidably :P) but undocumented. The example currently uses a Brütal shutdown (which was mostly a vestige from `tokio::io::stdin` using an uncancellable background worker task), but could be downgraded to a forceful shutdown.

The Kronblom and overcooked shutdowns would (IMO) mostly be useful for testing runtime internals, and this PR replaces those cases with sleeps (which are collapsed by tokio's testing mode anyway).

Reply:
Hah. That's a solid classification :metal:

Yeah, that sounds sensible. I think the graceful and the brutal ones are likely the most useful ones for us (maybe forceful as well, and as you say, overcooked ones for testing).

Everything in the PR so far looks sensible to me. But I'm still a bit unsure about the main example here:

If we are currently in a brutal scenario, what good would does the last `select!` in `main` do? If you removed the `process::exit`, and instead defer to the `forceful_shutdown` which i assume is intended to  trigger at the end of the async double-ctrl-c wait scope, then that's just immediately triggered in that last `select!` instead, right? i don't think it would functionally cause any different behaviour to avoid `process::exit`. Or am i misunderstanding.

Reply:
> i don't think it would functionally cause any different behaviour to avoid process::exit

`Controller` tries to abort all reconciliations when dropped, and `#[tokio::main]` waits for all spawned tasks to finish before exiting the process after the main function returns.

This combines to give you a forceful shutdown, rather than brutal (according to the previous chart :P).

> we are currently in a brutal scenario, what good would does the last select! in main do?

Regardless of whether we call `std::process::exit` in it, something needs to poll it for it to actually do anything. And we can't just `spawn` it, since that would keep the graceful shutdown waiting for it.

Reply:
Added a helper now for managing the signal handling for you.

Reply:
Ahhh. It's because futures does nothing unless polled. Bahhh. Sorry, I was being thick.

Reply:
Aren't we all?